broker.go

```diff
@@ -0,0 +1,604 @@
+// waiting goroutines to stop waiting on the channel.  For this reason, channel
+// readers should check for channel closure.
+func (lb *LegBroker) OnChange() (<-chan ChangeEvent, context.CancelFunc) {
+	// Channel is buffered to prevent distribute() from blocking if a reader is
+	// not reading the channel immediately.
+	ch := make(chan ChangeEvent, 1)
+	lb.outEventsMutex.Lock()
+	defer lb.outEventsMutex.Unlock()
+
+	lb.outEventsChans = append(lb.outEventsChans, ch)
+	cncl := func() {
+		lb.outEventsMutex.Lock()
+		defer lb.outEventsMutex.Unlock()
+		for i, ca := range lb.outEventsChans {
+			if ca == ch {
+				lb.outEventsChans[i] = lb.outEventsChans[len(lb.outEventsChans)-1]
+				lb.outEventsChans[len(lb.outEventsChans)-1] = nil
+				lb.outEventsChans = lb.outEventsChans[:len(lb.outEventsChans)-1]
+				close(ch)
+				break
+			}
+		}
+	}
+	return ch, cncl
+}
+
+// Sync performs a one-off explicit sync with the given peer for a specific CID and updates the latest
+// synced link to it.
+//
+// The Context passed in controls the lifetime of the background sync process
+// and not the call to sync, so make sure the context has an appropriate
+// deadline, if any.
+//
+// A change event channel is returned to allow the caller to wait for this particular sync
+// to complete.  Any OnChange watchers will also get a ChangeEvent when the sync succeeds.
+//
+// It is the responsibility of the caller to make sure the given CID appears after the latest sync
+// in order to avid re-syncing of content that may have previously been synced.
+//
+// The selector sequence, selSec, can optionally be specified to customize the
+// selection sequence during traversal.  If unspecified, the subscriber's
+// default selector sequence is used.
+//
+// Note that the selector sequence is wrapped with a selector logic that will stop traversal when
+// the latest synced link is reached. Therefore, it must only specify the selection sequence itself.
+//
+// See: ExploreRecursiveWithStopNode.
+func (lb *LegBroker) Sync(ctx context.Context, peerID peer.ID, c cid.Cid, selSeq ipld.Node) (<-chan ChangeEvent, error) {
+	if peerID == "" {
+		return nil, errors.New("empty peer id")
+	}
+
+	// Check for existing handler.  If none, create on if allowed.
+	hnd, err := lb.getOrCreateHandler(peerID, true)
+	if err != nil {
+		return nil, fmt.Errorf("cannot sync to peer: %s", err)
+	}
+
+	// Start goroutine to get latest root and handle
+	out := make(chan ChangeEvent, 1)
+	go func() {
+		defer close(out)
+
+		var updateLatest bool
+		if c == cid.Undef {
+			// Query the peer for the latest CID
+			var err error
+			c, err = head.QueryRootCid(ctx, lb.host, lb.topic.String(), peerID)
+			if err != nil {
+				log.Errorw("Cannot get root for sync", "err", err, "peer", peerID)
+				return
+			}
+			log.Debugw("Sync queried root CID", "peer", peerID, "cid", c)
+			if selSeq == nil {
+				// Update the latestSync only if no CID and no selector given.
```

Comment:
Copied from previous logic.  Here is I think the reasoning is:
- Do not update `latestSync` if given a specific CID, because it may not actually be the latest, and we may already have something later.
- Do not update `latestSync` if given a custom selector because who knows what it selects.